test/object-store/sync/flx_sync.cpp

```diff
@@ -4791,6 +4791,255 @@ TEST_CASE("flx: pause and resume bootstrapping at query version 0", "[sync][flx]
+                            REQUIRE(data.batch_state != BatchState::SteadyState);
+                            REQUIRE((curr_state == TestState::connected || curr_state == TestState::bs_downloading));
+                            if (data.batch_state == BatchState::LastInBatch) {
+                                next_state = TestState::bs_downloaded;
+                            }
+                            else {
+                                next_state = TestState::bs_downloading;
+                            }
+                            {
+                                std::lock_guard lock(callback_mutex);
+                                if (msg_count == 0) {
+                                    REQUIRE(last_state == BatchState::SteadyState);
+                                }
+                                else {
+                                    REQUIRE(last_state == BatchState::MoreToCome);
+                                }
+                                last_state = data.batch_state;
+                                ++msg_count;
+                            }
+                            return next_state;
+                        }
+                        case SyncClientHookEvent::BootstrapProcessed:
+                            REQUIRE(last_state == BatchState::LastInBatch);
+                            REQUIRE(curr_state == TestState::bs_downloaded);
+                            return TestState::bs_complete;
+                        case SyncClientHookEvent::DownloadMessageReceived:
+                        case SyncClientHookEvent::DownloadMessageIntegrated:
+                        default:
+                            return std::nullopt;
+                    }
+                });
+            return SyncClientHookAction::NoAction;
+        };
+        {
+            std::lock_guard lock(hook_mutex);
+            hook_callback = std::move(bs_callback);
+        }
+
+        // Updating the role to employees only
+        if (doc_filter) {
+            rule["roles"][0]["document_filters"]["read"] = *doc_filter;
+            rule["roles"][0]["document_filters"]["write"] = *doc_filter;
+        }
+        else {
+            rule["roles"][0]["document_filters"]["read"] = true;
+            rule["roles"][0]["document_filters"]["write"] = true;
+        }
+        app_session.admin_api.update_default_rule(app_session.server_app_id, rule);
+
+        // Client should receive an error and drop/reconnect the session
+        REQUIRE(machina.wait_for(TestState::disconnected));
+        REQUIRE(machina.wait_for(TestState::connected));
+
+        REQUIRE(!wait_for_download(*realm));
+        bool multi_msg = false;
+        {
+            std::lock_guard lock(callback_mutex);
+            if (msg_count > 1)
+                multi_msg = true;
+        }
+        // Verify a bootstrap occurred if multiple messages were received
+        REQUIRE((!multi_msg || machina.get() == TestState::bs_complete));
```

Comment:
don't we need to wait for the state machine to reach bs_complete before we try to get this state value? i'm also not sure what multi_msg is supposed to mean/do?

Reply:
Are we trying to assert here whether there should be multiple messages or just checking that if there happen to be multiple messages that they end in a complete bootstrap? Can some bootstraps in this test only have one message - how do we enforce that?